ilp.py

```diff
@@ -0,0 +1,187 @@
+    unique_edges = set(filtered_edges)
+
+    # Initialize the ILP problem
+    prob = pulp.LpProblem("TSOFencePlacement", pulp.LpMinimize)
+
+    # Define the decision variables for each unique edge
+    fences = {edge: pulp.LpVariable(f'f_{edge}', cat='Binary') for edge in unique_edges}
+
+    # Objective function: Minimize the number of fences
+    prob += pulp.lpSum(fences[edge] for edge in fences), "MinimizeFences"
+
+    # Constraints: Ensure each critical cycle is broken
+    for i, cycle in enumerate(cycles):
+        cycle_edges = []
+        for i in range(len(cycle)):
+            for j in range(len(cycle)):
+                if i != j:
+                    u = cycle[i]
+                    v = cycle[j]
+                    cycle_edges.append((u, v))
+        prob += pulp.lpSum(fences[edge] for edge in cycle_edges if edge in fences) >= 1, f"BreakCriticalCycle_{i}"
+    # Solve the problem
+    prob.solve()
+
+    # Prepare the results
+    result = {edge: int(pulp.value(fences[edge])) for edge in fences}
+
+    return result
+
+
+def find_relevant_edges(edges, cycles, node_weights, edge_weights):
+    relevant_edges = []
+    edge_index_map = {edge: i for i, edge in enumerate(edges)}
+
+    # Create adjacency list for ProgramOrder edges
+    program_order_graph = {}
+    for i, edge in enumerate(edges):
+        if edge_weights[i] == 'ProgramOrder':
+            if edge[0] not in program_order_graph:
+                program_order_graph[edge[0]] = []
+            program_order_graph[edge[0]].append(edge[1])
+
+    def has_program_order_path(u, v):
+        # Perform DFS to check for path from u to v using ProgramOrder edges
+        stack = [u]
+        visited = set()
+
+        while stack:
+            node = stack.pop()
+            if node == v:
+                return True
+            if node not in visited:
+                visited.add(node)
+                if node in program_order_graph:
+                    stack.extend(program_order_graph[node])
+        return False
+
+    for cycle in cycles:
+        for i in range(len(cycle)):
+            for j in range(len(cycle)):
+                if i != j:
+                    u = cycle[i]
+                    v = cycle[j]
+                    edge = (u, v)
+                    if edge in edge_index_map:
+                        edge_index = edge_index_map[edge]
+                        if (node_weights[u][0] == 'Write' and node_weights[v][0] == 'Read' and edge_weights[
+                            edge_index] == 'ProgramOrder'):
+                            relevant_edges.append(edge)
+                    elif has_program_order_path(u, v):
+                        if node_weights[u][0] == 'Write' and node_weights[v][0] == 'Read':
+                            relevant_edges.append(edge)
+    return relevant_edges
+
+
+if __name__ == "__main__":
+    # Example usage
+    edges = [(0, 1), (2, 3), (4, 5), (5, 6), (6, 7), (8, 9), (10, 11), (0, 7), (7, 0),
```

Comment:
as Nicolae mentioned, it might be useful to have this information stored in a class. right now I'm using this for the ALNS part:

```
class AegInstance:
    def __init__(self, edges, cycles, node_weights, edge_weights):
        self.edges = edges
        self.cycles = cycles
        self.node_weights = node_weights
        self.edge_weights = edge_weights
```

which could be expanded with the functions above